src/compiler/utilities.ts

```diff
@@ -2726,8 +2726,9 @@ namespace ts {
         return forEach(supportedTypeScriptExtensions, extension => fileExtensionIs(fileName, extension));
     }
 
-    export function hasTypeScriptFileExtensionNonDts(fileName: string) {
-        return forEach(supportedTypeScriptExtensionsNonDts, extension => fileExtensionIs(fileName, extension));
+    /** Return ".ts" or ".tsx" if that is the extension. */
+    export function tryExtractTypeScriptExtensionNonDts(fileName: string): string | undefined {
+        return find(supportedTypeScriptExtensionsNonDts, extension => fileExtensionIs(fileName, extension));

```

Comment:
> If people add a .d.ts extension they may be doing unnecessary work, but it won't cause runtime errors because declaration references don't translate to imports in the output.

`.d.ts` imports don't just get elided. If you then use any imported entities as values, then we still emit the import itself. Try it out:

`foo.d.ts`

```ts
export declare var foo;
```

`bar.ts`
```ts
import * as v from "./foo.d.ts";

v.foo;
```

The current emit for `bar.js` is

```js
"use strict";
var v = require("./foo.d.ts");
v.foo;
```